### ISSUE TITLE
resizing panes is very slow under unclear circumstances

### ISSUE THREAD

**Issue by benaryorg** [tmux/tmux#4171]
### Issue description

First of all, sorry for the vague issue, I've been dealing with this issue for quite a while now and only a few hours ago actually started to look into it more closely.
What makes this so aggravating is that I can't cleanly reproduce it, and it only happens after I've had tmux open for a few days or weeks, but then it is very consistent.
I'll describe my setup first, and then describe what I've tried so far.

Basically I have three panes open, one of which is vim (which runs on the alternate buffer so there's practically no history in that pane) and the two others being actively used for doing things and accumulating history over the span of days to weeks.
All panes have a history-limit of 50,000 lines.
There are several other windows in the session (and another session), but those don't really seem to make a difference; if I have any other window selected, i.e. one that hasn't seen much use or is freshly created, then the panes work smoothly in every respect.

The issue itself manifests whenever resizing of the panes happens, either because the terminal on which the tmux client is running is resized (due to the terminal emulator resizing), or because I try to resize the panes manually.
Whenever that happens, but only ever when tmux has been running for a while and the panes used somewhat, then resizing takes a long time, anything from 2-3 seconds to upwards of 15 seconds, which is really annoying on a tiling window manager because whenever I code something and have to take a look at docs or something my entire workflow freezes on the spot together with tmux.

What does fix the issue ad-hoc is to clear the history of all involved panes (`clearhist`).

I'm pretty sure this issue surfaced when I moved to 3.4, and hasn't bothered me (at least not to that extent) on 3.3a.
Because it's been somewhat sporadic I couldn't think of a good way to reproduce and today when I started looking hit that exact issue; when trying to deliberately cause this behaviour I couldn't; 3.3a, 3.4, and 486221f2063e211cb900e7f2853645b676f6e963 all behaved within usual latency (i.e. resizing taking like half a second after filling four panes on the same window with 50k lines of `base64 -w 200 /dev/urandom`).
There is a point to be made here that resizing shouldn't take that long anyway since tmux could *in theory* (I realize that this may be incredibly roundabout to implement) try to scan from the end of the buffer up, so on resize it wouldn't have to render the entire text, but given that curses and escape sequences are a thing that can stretch multiline it may not be that easy of course.
The one thing that stuck out however in the different use cases are: a) I occasionally have applications rendering updating multiline status output (colmena or nix do this) but I'm pretty sure I've had issues without either of those before but I'm not 100% on that, and b) tmux has introduced the timestamps within the scrollback buffer since and I think I remember the issue first appearing when those timestamps came around.

I can't really hand out logs of that tmux server because it's the one I use daily and that'd not just mean that the logs would be tens to hundreds of MiB large (which probably doesn't help much?), but also that they contain data that may be sensitive so I'm wary of providing that.

I did however run flamegraph against the tmux client and server, and while the server seems pretty relaxed (basically being in a sleep state during the freeze), the client does show this:

![flamegraph](https://github.com/user-attachments/assets/dbd80266-ca9e-46e1-9548-47a6cb96c363)

(Note that this is my currently running 3.4 tmux, probably compiled with `-O2` or something)

This makes a strong case for some reallocation happening on resizing that maybe does not need to happen, but as someone who doesn't know the codebase I wouldn't know for certain.

Some closing words; I would *really* love for this issue/performance regression/whatever to be resolved because as someone whose terminal keybind literally starts `alacritty -e tmux new-session -t work -f active-pane` and who basically spends 50% of their time in the shell this is really annoying.
I can try to upgrade to the master branch, however it'll probably take a while for any of the issues to resurface then, so I'd much rather get some advice on what debugging steps would help the most right now.
I assume having tmux from *master* compiled with `-O0` would be ideal, anything else that's relevant?
That is, if you can't already magically discern where the bottleneck lies from the information above (I'd be happy to test any patches of course).

### Required information

Please provide the following information:

* tmux version: 3.4
* Platform: Linux unknown
* $TERM inside and outside of tmux: alacritty ⇒ tmux
* Logs from tmux: see description above, could provide if deemed necessary

**Comment by nicm** [comment]
You must have the client and server the wrong way round because only the server will ever call `grid_reflow_join`. This is tmux putting lines together that should not be wrapped for the new width.

Unless you can come up with a way to reproduce this I don't think there is much we can do here however.

**Comment by nicm** [comment]
When it happens, you could go to all the open panes and run `capture-pane -pJe -S- -E-` them (you will need to do it from another window), then you can try to `cat` that output in a new server and see if that shows the same behaviour.

**Comment by nicm** [comment]
Are you sure `-w 200` is enough? Did you try different variations of line lengths?

**Comment by benaryorg** [comment]
> You must have the client and server the wrong way round because only the server will ever call `grid_reflow_join`. This is tmux putting lines together that should not be wrapped for the new width.

Ah, yeah, my mistake, the server freezes while the client idles.
Somehow I thought that the server must have the lower pid since the client connects to the server but now that I think about it that doesn't make sense so I seem to have mixed them up there.

> Are you sure `-w 200` is enough? Did you try different variations of line lengths?

200 characters is enough for it to be split onto two lines for two panes side-by-side on my setup so I could test how the transition from one-per-line to multiline went (as expected it hangs hard when crossing that threshold, but not as hard as to explain the behaviour I have in the other session).
However given the next part of this reply I think the issue lies less with the length and more the underline style?

> When it happens, you could go to all the open panes and run `capture-pane -pJe -S- -E-` them (you will need to do it from another window), then you can try to `cat` that output in a new server and see if that shows the same behaviour.

Thank you very much, I should've thought of that *\*sigh\**
I can now confirm that *master* is fixed apparently.
Two git bisects further I can tell you that the issue was apparently introduced with 828efc7bcf721049c85f108a9cebe798079cbd06 and fixed by c773fe89e7ac75fbf86bfce30d86ebf44e5c20e2.
Given that both of these patches deal with a replacement of 0 by 8 or vice versa I feel like this actually was the culprit, but honestly I have no idea why.

<details>

<summary>Bisect logs for the sake of it.</summary>

Here's the bisect for when the issue appeared:

```text
# broken: [9ae69c3795ab5ef6b4d760f6398cd9281151f632] 3.4.
# fine: [0b355ae8114511e1ff6359272b164f1cdf718e80] 3.3a.
git bisect start '--first-parent' '--term-old=fine' '--term-new=broken' '3.4' '3.3a'
# fine: [023b0f76c3712281e30041f086b91c89f9525d33] Merge branch 'obsd-master'
git bisect fine 023b0f76c3712281e30041f086b91c89f9525d33
# broken: [dfbc6b1888c110cf0ade66f20188c57757ee1298] Merge SIXEL branch.
git bisect broken dfbc6b1888c110cf0ade66f20188c57757ee1298
# fine: [237ee6f2311b351b8a3bb3b46e17622f6e3b6bb9] Merge branch 'obsd-master'
git bisect fine 237ee6f2311b351b8a3bb3b46e17622f6e3b6bb9
# broken: [828efc7bcf721049c85f108a9cebe798079cbd06] Merge branch 'obsd-master'
git bisect broken 828efc7bcf721049c85f108a9cebe798079cbd06
# fine: [e4c4ceb2865fd9d2006a85c7f48bb7b6dd2f5c49] Merge branch 'obsd-master'
git bisect fine e4c4ceb2865fd9d2006a85c7f48bb7b6dd2f5c49
# fine: [18870913c5325988a7eb61cce4af09576da3fd45] Merge branch 'obsd-master'
git bisect fine 18870913c5325988a7eb61cce4af09576da3fd45
# fine: [c8494dff7b6b9a996866edaf8c57c4e4b393acd0] Merge branch 'obsd-master'
git bisect fine c8494dff7b6b9a996866edaf8c57c4e4b393acd0
# first broken commit: [828efc7bcf721049c85f108a9cebe798079cbd06] Merge branch 'obsd-master'
```

and when it disappeared:

```text
# fixed: [5b7bdc82c8c87d04309d71643a951a7c1aceddce] Merge branch 'obsd-master'
# broken: [9ae69c3795ab5ef6b4d760f6398cd9281151f632] 3.4.
git bisect start '--first-parent' '--term-new=fixed' '--term-old=broken' 'master' '3.4'
# fixed: [2ac0faf119525ef6d2c4f8f1875c3f51ec224d33] If built with systemd, remove some environment variables it uses. From Ciprian Dorin Craciun, GitHub issue 4035.
git bisect fixed 2ac0faf119525ef6d2c4f8f1875c3f51ec224d33
# broken: [9ba433e521434abdd82babec1041d64b5f5c94ef] Use printf not echo -e, from Joyce Lin.
git bisect broken 9ba433e521434abdd82babec1041d64b5f5c94ef
# broken: [4c2eedca5a75dedc3540e3373048ce7979c82064] Merge branch 'obsd-master'
git bisect broken 4c2eedca5a75dedc3540e3373048ce7979c82064
# fixed: [d02254f754e0a116712abf30b54ca0d4a82216a3] Merge branch 'obsd-master'
git bisect fixed d02254f754e0a116712abf30b54ca0d4a82216a3
# fixed: [c773fe89e7ac75fbf86bfce30d86ebf44e5c20e2] Merge branch 'obsd-master'
git bisect fixed c773fe89e7ac75fbf86bfce30d86ebf44e5c20e2
# broken: [c07e856d244d07ab2b65e72328fb9fe20747794b] Merge branch 'obsd-master'
git bisect broken c07e856d244d07ab2b65e72328fb9fe20747794b
# first fixed commit: [c773fe89e7ac75fbf86bfce30d86ebf44e5c20e2] Merge branch 'obsd-master'
```

</details>

> Unless you can come up with a way to reproduce this I don't think there is much we can do here however.

Thanks for pointing me in the capture-pane direction, somehow I completely missed that I could try that.
I also got to use *git bisect* again which is a tool I dearly love and get to use way too little these days.
Anyway, I guess I'll close the issue here as it seems that I'll live in a world of having a fixed tmux with the next release of my distro in about a month.
Also sorry for bothering you with an issue that was already fixed.

**Comment by nicm** [comment]
I reckon it was probably fixed by db1665868ff33375f6ea235ffbbb1996d7946c05 which fixed tmux treating every cell as an extended cell (much more costly).
